### ISSUE TITLE
Error: rule quantify output/rule draw_heatmaps 

### ISSUE THREAD

**Issue by thierryjanssens** [DennisSchmitz/Jovian_archive#41]
When using Jovian v0.9.2-1-g5668474 the combine failure of above mentioned rules remains.

**Terminal output:**

```
Activating conda environment: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0

Now counting the number of lines in the fastq files from which
reads have been discarded that were mapped to the human genome
by Bowtie2 to determine the number of non-human reads.

This may take a while...

Done counting!===============================================] 100.0% - Reading files  [ 39 / 39 ]
Traceback (most recent call last):
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/scripts/tmp9xe33yh_.quantify_profiles.py", line 487, in <module>
    main()
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/scripts/tmp9xe33yh_.quantify_profiles.py", line 421, in main
    "Eukaryota", "Viruses", "Unclassified" ]]
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0/lib/python3.7/site-packages/pandas/core/frame.py", line 2682, in __getitem__
    return self._getitem_array(key)
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0/lib/python3.7/site-packages/pandas/core/frame.py", line 2726, in _getitem_array
    indexer = self.loc._convert_to_indexer(key, axis=1)
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0/lib/python3.7/site-packages/pandas/core/indexing.py", line 1327, in _convert_to_indexer
    .format(mask=objarr[mask]))
KeyError: "['Archaea'] not in index"
[Fri May 10 15:11:00 2019]
Error in rule quantify_output:
    jobid: 134
    output: results/profile_read_counts.csv, results/profile_percentages.csv, results/Sample_composition_graph.html
    log: logs/quantify_output.log
    conda-env: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0

RuleException:
CalledProcessError in line 633 of /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/Snakefile:
Command 'source /mnt/miniconda/bin/activate '/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0'; set -euo pipefail;  python /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_
06MAY2019/Jovian/.snakemake/scripts/tmp9xe33yh_.quantify_profiles.py' returned non-zero exit status 1.
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/Snakefile", line 633, in __rule_quantify_output
  File "/home/janssetk/.conda/envs/Jovian_master/lib/python3.6/concurrent/futures/thread.py", line 56, in run
[Fri May 10 15:11:01 2019]
Error in rule draw_heatmaps:
    jobid: 133
    output: results/Superkingdoms_quantities_per_sample.csv, results/heatmaps/Superkingdoms_heatmap.html, results/heatmaps/Virus_heatmap-order.html, results/heatmaps/Virus_heatmap-family.html, results/heatmaps/Virus_heatmap-genus.html, r
esults/heatmaps/Virus_heatmap-species.html, results/heatmaps/Phage_heatmap-order.html, results/heatmaps/Phage_heatmap-family.html, results/heatmaps/Phage_heatmap-genus.html, results/heatmaps/Phage_heatmap-species.html, results/heatmaps/B
acteria_heatmap-phylum.html, results/heatmaps/Bacteria_heatmap-class.html, results/heatmaps/Bacteria_heatmap-order.html, results/heatmaps/Bacteria_heatmap-family.html, results/heatmaps/Bacteria_heatmap-genus.html, results/heatmaps/Bacter
ia_heatmap-species.html, results/Taxonomic_rank_statistics.tsv, results/Virus_rank_statistics.tsv, results/Phage_rank_statistics.tsv, results/Bacteria_rank_statistics.tsv
    log: logs/draw_heatmaps.log
    conda-env: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0

ClusterJobException in line 639 of /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/Snakefile:
Error executing rule draw_heatmaps on cluster (jobid: 133, external: 268289, jobscript: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/tmp.5w1wg5n7/Jovian_draw_heatmaps.jobid133). For detailed err
or see the cluster log.
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/log/2019-05-10T110032.969275.snakemake.log
```

**DRMAA output:**

```
Sender: LSF System <lsfadmin@rivm-lsfsv-l05p.rivm.ssc-campus.nl>
Subject: Job 268289: <Jovian_draw_heatmaps.jobid133> in cluster <RIVM-LSF-PRD> Exited

Job <Jovian_draw_heatmaps.jobid133> was submitted from host <rivm-biohn-l01p.rivm.ssc-campus.nl> by user <janssetk> in cluster <RIVM-LSF-PRD> at Fri May 10 15:10:31 2019.
Job was executed on host(s) <rivm-lsfsv-l05p.rivm.ssc-campus.nl>, in queue <bio>, as user <janssetk> in cluster <RIVM-LSF-PRD> at Fri May 10 15:10:32 2019.
</home/janssetk> was used as the home directory.
</data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian> was used as the working directory.
Started at Fri May 10 15:10:32 2019.
Terminated at Fri May 10 15:10:50 2019.
Results reported at Fri May 10 15:10:50 2019.

Your job looked like:

------------------------------------------------------------
# LSBATCH: User input
/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/tmp.5w1wg5n7/Jovian_draw_heatmaps.jobid133
------------------------------------------------------------

Exited with exit code 1.

Resource usage summary:

    CPU time :                                   3.29 sec.
    Max Memory :                                 47 MB
    Average Memory :                             34.00 MB
    Total Requested Memory :                     -
    Delta Memory :                               -
    Max Swap :                                   -
    Max Processes :                              6
    Max Threads :                                9
    Run time :                                   24 sec.
    Turnaround time :                            19 sec.

The output (if any) follows:

Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 1
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	draw_heatmaps
	1

[Fri May 10 15:10:38 2019]
rule draw_heatmaps:
    input: results/all_taxClassified.tsv, results/multiqc_data/multiqc_trimmomatic.txt
    output: results/Superkingdoms_quantities_per_sample.csv, results/heatmaps/Superkingdoms_heatmap.html, results/heatmaps/Virus_heatmap-order.html, results/heatmaps/Virus_heatmap-family.html, results/heatmaps/Virus_heatmap-genus.html, results/heatmaps/Virus_heatmap-species.html, results/heatmaps/Phage_heatmap-order.html, results/heatmaps/Phage_heatmap-family.html, results/heatmaps/Phage_heatmap-genus.html, results/heatmaps/Phage_heatmap-species.html, results/heatmaps/Bacteria_heatmap-phylum.html, results/heatmaps/Bacteria_heatmap-class.html, results/heatmaps/Bacteria_heatmap-order.html, results/heatmaps/Bacteria_heatmap-family.html, results/heatmaps/Bacteria_heatmap-genus.html, results/heatmaps/Bacteria_heatmap-species.html, results/Taxonomic_rank_statistics.tsv, results/Virus_rank_statistics.tsv, results/Phage_rank_statistics.tsv, results/Bacteria_rank_statistics.tsv
    log: logs/draw_heatmaps.log
    jobid: 0
    benchmark: logs/benchmark/draw_heatmaps.txt

Activating conda environment: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0
[Fri May 10 15:10:50 2019]
Error in rule draw_heatmaps:
    jobid: 0
    output: results/Superkingdoms_quantities_per_sample.csv, results/heatmaps/Superkingdoms_heatmap.html, results/heatmaps/Virus_heatmap-order.html, results/heatmaps/Virus_heatmap-family.html, results/heatmaps/Virus_heatmap-genus.html, results/heatmaps/Virus_heatmap-species.html, results/heatmaps/Phage_heatmap-order.html, results/heatmaps/Phage_heatmap-family.html, results/heatmaps/Phage_heatmap-genus.html, results/heatmaps/Phage_heatmap-species.html, results/heatmaps/Bacteria_heatmap-phylum.html, results/heatmaps/Bacteria_heatmap-class.html, results/heatmaps/Bacteria_heatmap-order.html, results/heatmaps/Bacteria_heatmap-family.html, results/heatmaps/Bacteria_heatmap-genus.html, results/heatmaps/Bacteria_heatmap-species.html, results/Taxonomic_rank_statistics.tsv, results/Virus_rank_statistics.tsv, results/Phage_rank_statistics.tsv, results/Bacteria_rank_statistics.tsv
    log: logs/draw_heatmaps.log
    conda-env: /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0

RuleException:
CalledProcessError in line 665 of /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/Snakefile:
Command 'source /mnt/miniconda/bin/activate '/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/.snakemake/conda/f4fd10e0'; set -euo pipefail;  python bin/draw_heatmaps.py -c results/all_taxClassified.tsv -n results/multiqc_data/multiqc_trimmomatic.txt -sq results/Superkingdoms_quantities_per_sample.csv -st results/Taxonomic_rank_statistics.tsv -vs results/Virus_rank_statistics.tsv -ps results/Phage_rank_statistics.tsv -bs results/Bacteria_rank_statistics.tsv -s results/heatmaps/Superkingdoms_heatmap.html -v results/heatmaps/Virus_heatmap.html -p results/heatmaps/Phage_heatmap.html -b results/heatmaps/Bacteria_heatmap.html > logs/draw_heatmaps.log 2>&1' returned non-zero exit status 1.
  File "/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/Snakefile", line 665, in __rule_draw_heatmaps
  File "/home/janssetk/.conda/envs/Jovian_master/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Removing output files of failed job draw_heatmaps since they might be corrupted:
results/Superkingdoms_quantities_per_sample.csv, results/heatmaps/Superkingdoms_heatmap.html, results/heatmaps/Virus_heatmap-order.html, results/heatmaps/Virus_heatmap-family.html, results/heatmaps/Virus_heatmap-genus.html, results/Taxonomic_rank_statistics.tsv, results/Virus_rank_statistics.tsv, results/Phage_rank_statistics.tsv, results/Bacteria_rank_statistics.tsv
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message


PS:

Unable to read stderr data from stderr buffer file; your job was probably aborted prematurely.
```

**Comment by DennisSchmitz** [comment]
> KeyError: "['Archaea'] not in index"

I thought this was fixed. Or is this now the same bug in the `quantify_output` rule?

@samnooij can you please look into it.

**Comment by samnooij** [comment]
> > KeyError: "['Archaea'] not in index"
> 
> I thought this was fixed. Or is this now the same bug in the `quantify_output` rule?
> 
> @samnooij can you please look into it.

Indeed, it seems I forgot that `quantify_output` assumes presence of all superkingdoms too... I will try and apply the same fix as with `draw_heatmaps`.
About the second error: I am not sure yet what is going on there. @thierryjanssens  Could it be that there are no phages or bacteria in the sample? And does the rule `draw_heatmaps` in this case still create a log file? If so, does it say anything suspicious? (It should be in `logs/draw_heatmaps.log`.)

P.S. If it is the case that there are no phages and bacteria in the sample, and that the script fails because of that, I guess that would increase the priority of enabling users to select clades/taxa of interest for making heatmaps (issue #38).

**Comment by thierryjanssens** [comment]
@samnooij Thank you for your swift reply.
Rule draw_heatmaps also returns an error and an empty log file. There are indeed bacteria and phages in the samples, according to the lca output (as observed in the Krona diagrams).

**Comment by samnooij** [comment]
I have tried running the draw_heatmaps script in the directory you were working in (`/data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian/`) and learned a few things:  
1. the heatmap function is not ready for empty data. I.e. if there are no taxa in the clade that it is going to plot, the script will fail. In this case, there were no viral species, so the pipeline got stuck there.
2. By filling in zeroes in otherwise empty (NaN) fields, the heatmap will fill in every single block with data and colour them in a very light shade, whereas before these fields were left empty/blank. I prefer the empty fields as that is the easiest way to tell the absence of certain taxa, I think.

So `draw_heatmaps` will require (at least) 2 fixes:
**1. Check if dataframe is empty, and if so, better write an empty file and report in the log file. --PRIORITY**
2. Disable plotting for fields with 0 contigs. (Change 0 values back to NaN before plotting, I guess.)

As for the `quantify_profiles`: I have not yet checked it. I hope it will be an easy fix with 'add a column of zeroes for each missing superkingdom'. I will report back later!

**Comment by samnooij** [comment]
I have proposed a bugfix for the script, which seems to work.
(branch = quantify_profiles, command = `bash jovian -i raw_data/ -p results/profile_read_counts.csv`, directory = /data/BioGrid/ERVINGS/Runs_Respiratory_MiSEQ_RIVM/MiSeq_RUN_06MAY2019/Jovian)
Snakemake complains about 'Mising metadata for file ...' for rules that were not run, but otherwise it works. No errors, it finishes and creates the output files.
However, the plot in `results/Sample_composition_graph.html` again makes >100 bars in the percentage plot. My guess is that that is due to the sorting of sample names: it seems that I sorted them starting with S10, then S11, S12, and then S1, S2, etc. And S0 is at the end. Any ideas how to sort those properly?

Anyway, I guess it is best for now to accept this fix and work from there.
